### PR TITLE
[CS2113T-W12-1] FitNUS Tracker

### PR DESCRIPTION
**Pull request by kengjit**
**FitNUS Tracker** is a desktop app for tracking daily food intake to keep a healthy lifestyle. It is based on a Command Line Interface (CLI) and has a rich function list in order to cater to the needs of potential users.

### CODE REVIEW

File: docs/DeveloperGuide.md [hunk 1]
```diff
@@ -1,38 +1,450 @@
...
+
 ### Target user profile
 
-{Describe the target user profile}
+NUS Computer Engineering students reside in UTown going on diet.
 
 ### Value proposition
 
-{Describe the value proposition: what problem does it solve?}
+Help user to keep track of their daily calorie intake, and manage their diet wisely.
```
Comment (ng-andre) [inline docs/DeveloperGuide.md]: User profile and value proposition are quite brief, can consider elaborating on it more.

File: docs/DeveloperGuide.md [hunk 2]
```diff
@@ -1,38 +1,450 @@
...
+
+Refer to the User Guide (no link for now) for details of each command.
 
 ## User Stories
 
-|Version| As a ... | I want to ... | So that I can ...|
-|--------|----------|---------------|------------------|
-|v1.0|new user|see usage instructions|refer to them when I forget how to use the application|
-|v2.0|user|find a to-do item by name|locate a to-do without having to go through the entire list|
+|Version| As a ... | I want to ... | So that I can ...| |--------|----------|---------------|------------------| |v1.0|new user|see usage instructions|refer to them when I forget how to use the application| |v2.0|user|find a to-do item by name|locate a to-do without having to go through the entire list|
```
Comment (ng-andre) [inline docs/DeveloperGuide.md]: User stories formatting is broken and does not render correctly on the site.
image

File: docs/DeveloperGuide.md [hunk 3]
```diff
@@ -1,38 +1,450 @@
...
+- `User`: For handling all functionality regarding personalisation of user experience.
+
+####How the overall architecture works
+
+1. When the user enters a command, `FitNUS` uses the Parser class to parse the user command.
+2. This results in a `Command` object (more precisely, an object of one of its subclasses e.g., `AddFoodEntryCommand`). 
+3. The `Command` object calls its `execute` method which performs the function required.
+   - Since the `execute` method receives the `FoodDatabase`, `EntryDatabase` and `MealPlanDatabase` initialised in `FitNUS`, it is able to perform operations related to those components (e.g. to add a food tracker entry).
+4. The `execute` method returns a `String` object that contains the outcome message of the command that was executed, which is displayed to the user by the `Ui` component.
+---
```
Comment (ng-andre) [inline docs/DeveloperGuide.md]: Formatting is broken and/or inconsistent for this block, might want to fix the markdown.
image

File: docs/DeveloperGuide.md [hunk 4]
```diff
@@ -1,38 +1,450 @@
...
+
+![command class diagram](diagrams/command%20class%20diagram.drawio.png)
+
+The `Command` class is an abstract class that all other specific command classes (eg AddFoodEntryCommand, DeleteEntryCommand) inherit from. 
+
+The `Command` component
+    
+- Contains an abstract method `execute`. In the specific command classes that inherit from `Command`, `execute` performs the function that the command describes. (For example, in `AddFoodEntryCommand`, `execute` adds a food tracker entry to the food tracker.) 
+
+####Add Food Entry Feature
```
Comment (ng-andre) [inline docs/DeveloperGuide.md]: Formatting is broken again, same for `####Edit Food Entry Feature` and `####List Food Entry Feature` below

File: docs/DeveloperGuide.md [hunk 5]
```diff
@@ -1,38 +1,450 @@
...
+
+> ⚠️ Notes about the generate and set calorie goal feature:
+>- The weekly change is the absolute value of the weekly change in weight. It cannot be greater than 1.0, which is the upper bound for the recommended healthy weight change per week.
+
+#### Setting user data feature
+
+The user is able to change their personal data at any point while using the app. 
+
+Setting gender, age and height operate in a similar way, as shown in the example sequence diagram below where setting height is performed:
+![SetHeightSeqDiagram](./diagrams/SetHeightCommand.png "Set Height Sequence Diagram")
```
Comment (ng-andre) [inline docs/DeveloperGuide.md]: UML Diagram should be on a new line for more readability.

File: docs/DeveloperGuide.md [hunk 6]
```diff
@@ -1,38 +1,450 @@
...
+*   `generateMonthSummaryReport()` shows average daily calorie intake, and the most/least frequently eaten food over this month.
+
+#### UML Sequence Diagram
+
+The following sequence diagram describes the operation of the `generateWeekSummary()`.  
+![](diagrams/weekly-report.png)  
+The following sequence diagram describes the operation of `generateMonthSummary()`.  
+![](diagrams/monthly-report.png)
+
+---
```
Comment (ng-andre) [inline docs/DeveloperGuide.md]: The two sequence diagrams are inconsistent with the rest of the diagrams in terms of formatting and colour. For the `generateWeekSummary()` sequence diagram, it is slightly unclear as the entity name is `:viewMonthSummaryCommand` and might be potentially misleading.

File: docs/DeveloperGuide.md [hunk 7]
```diff
@@ -1,38 +1,450 @@
...
+- Stores the user's personal data eg gender, age, height, weight
+- Stores the user's weight progress data i.e. all `WeightProgressEntry` objects (which are contained in an `ArrayList` as an attribute in `User`)
+- Performs functions related to the user's calorie goal such as setting and generating the calorie goal
+
+#### Weight tracker feature
+
+The weight tracker consists of the `ArrayList` of `WeightProgressEntry` objects. Each `WeightProgressEntry` object stores a date as a `LocalDate` and the weight corresponding to the date stored.
+
+The `updateWeightAndWeightTracker` method allows the user to update their weight and the weight tracker. This is performed as shown in the following sequence diagram:
+![SetWeightSeqDiagram](./diagrams/SetWeightCommand.png "Set Weight Sequence Diagram")
```
Comment (ng-andre) [inline docs/DeveloperGuide.md]: The sequence diagrams for `Weight tracker feature`, `Calories remaining feature` and `Generate and set calorie goal feature` may be unnecessarily repetitive as they have the same overall strucuture, with the only differences being the entity names and method call names. Can consider using only one example.

File: docs/DeveloperGuide.md [hunk 8]
```diff
@@ -1,38 +1,450 @@
-# Developer Guide
+#Developer Guide
 
-## Acknowledgements
+##Content
```
Comment (kairoskoh) [inline docs/DeveloperGuide.md]: Is the use of # intentional or a syntax error? It shows up with the # in the developer guide. It might be better to have a uniform format throughout the document. To fix this, simply leave a space between the # and text.
image

File: docs/DeveloperGuide.md [hunk 9]
```diff
@@ -1,38 +1,450 @@
...
-|v1.0|new user|see usage instructions|refer to them when I forget how to use the application|
-|v2.0|user|find a to-do item by name|locate a to-do without having to go through the entire list|
+|Version| As a ... | I want to ... | So that I can ...| |--------|----------|---------------|------------------| |v1.0|new user|see usage instructions|refer to them when I forget how to use the application| |v2.0|user|find a to-do item by name|locate a to-do without having to go through the entire list|
+
+## Architecture
+
+### Overall Architecture
+
+![Overall Architecture Diagram](diagrams/overall%20architecture.png)  
+The Architecture Diagram given above explains the high-level design of the App.
```
Comment (kairoskoh) [inline docs/DeveloperGuide.md:77]: Would it be clearer if the user and database/storage is labelled clearly?

File: docs/DeveloperGuide.md [hunk 10]
```diff
@@ -1,38 +1,450 @@
...
+   - [User](###User)
+   - [Summary](###Summary)
+   - [Suggest](###Suggest)
+   - [Command](###Command)
+   - [Storage](###Storage)
+   - [Parser](###Parser)
+   - [Ui](###Ui)
+5. [Implementation](#Implementation)
+6. [Instruction for manual testing](#instruction-for-manual-testing)
+7. [Non-functional Requirement](#NF-Requirement)
```
Comment (kairoskoh) [inline docs/DeveloperGuide.md]: Would the use of numbering indexes later on in the developer guide be helpful?
The numbering in the contents page is great, but it doesnt continue on later down in the guide.
The guide appears to be a long chunk of text and is hard to follow.

File: docs/DeveloperGuide.md [hunk 11]
```diff
@@ -1,38 +1,450 @@
...
 
-## Acknowledgements
+##Content
 
-{list here sources of all reused/adapted ideas, code, documentation, and third-party libraries -- include links to the original source as well}
 
-## Design & implementation
+1. [Product Scope](##Product-Scope)
+2. [Quick Start](##quick-start)
+3. [User Story](##user-story)
```
Comment (kairoskoh) [inline docs/DeveloperGuide.md]: Would it be better to have working hyperlinks in the contents section?
Some of the hyperlinks are not working. Only 6. Instruction for manual testing worked when I tested it.

File: docs/DeveloperGuide.md [hunk 12]
```diff
@@ -1,38 +1,450 @@
...
+based on the remaining calories and specified type. The user also has the option to have the result sorted 
+in ascending order of calories. This is indicated by the boolean `isSort` variable.
+4. The returned ArrayList of matching `Food` objects is passed to `Ui` to be printed to the user.
+
+---
+
+### Command
+
+![command class diagram](diagrams/command%20class%20diagram.drawio.png)
+
```
Comment (kairoskoh) [inline docs/DeveloperGuide.md:461]: Is there a better way to represent this diagram? All the different command classes inherits from a single class. The diagram is very long right now.

File: docs/DeveloperGuide.md [hunk 13]
```diff
@@ -1,38 +1,451 @@
-# Developer Guide
+#Developer Guide
 
-## Acknowledgements
+##Content
```
Comment (jonathanmui4) [inline docs/DeveloperGuide.md]: I agree with Kairos. Removing the space between the `#` and the header causes an error that results in the header not being registered as one in Github pages. Perhaps, you can consider standardising the formatting of the headers and leaving a space between the `#` and the header name in all subsequent headers i.e. `# Developer Guide`.

File: docs/DeveloperGuide.md [hunk 14]
```diff
@@ -1,38 +1,451 @@
...
+   - [User](###User)
+   - [Summary](###Summary)
+   - [Suggest](###Suggest)
+   - [Command](###Command)
+   - [Storage](###Storage)
+   - [Parser](###Parser)
+   - [Ui](###Ui)
+5. [Implementation](#Implementation)
+6. [Instruction for manual testing](#instruction-for-manual-testing)
+7. [Non-functional Requirement](#NF-Requirement)
```
Comment (jonathanmui4) [inline docs/DeveloperGuide.md]: Perhaps you can consider the following points to fix the hyperlinks in your Developer Guide:

1. Only 1 `#` is required in the link segment (text in the rounded brackets) regardless of the header type
2. Link segment is lower case
3. Spaces are replaced with `-`

i.e. ` [Overall Architecture](#overall-architecture)`

File: docs/DeveloperGuide.md [hunk 15]
```diff
@@ -1,38 +1,451 @@
...
+## Acknowledgements
 
+{list here sources of all reused/adapted ideas, code, documentation, and third-party libraries -- include links to the original source as well}
 
 ## Product scope
+
 ### Target user profile
 
-{Describe the target user profile}
+NUS Computer Engineering students reside in UTown going on diet.
```
Comment (jonathanmui4) [inline docs/DeveloperGuide.md]: Perhaps, you can consider introducing the problem you are trying to solve for your target users in your "target user profile" too. 

E.g. Has a need plan and organise their diets quickly, comfortable with CLI apps, can type fast etc.

File: docs/DeveloperGuide.md [hunk 16]
```diff
@@ -1,38 +1,451 @@
-# Developer Guide
+#Developer Guide
```
Comment (IshaaanVyas) [inline docs/DeveloperGuide.md]: The # and ## are showing on the DG page

File: docs/DeveloperGuide.md [hunk 17]
```diff
@@ -1,38 +1,451 @@
...
+#### Generate and set calorie goal feature
+
+The generate and set calorie goal feature generates a calorie goal according to the user's desired weekly weight change, age, height, weight and gender, and then sets the user's calorie goal to the generated goal. 
+
+This is performed as shown in the following sequence diagram:
+![GenerateGoalSeqDiagram](./diagrams/GenerateCalorieGoalCommand.png "Generate Calorie Goal Sequence Diagram")
+
+- The following formulas are used to generate the calorie goal:
+  - For females: calorieGoal = [[655.1 + (9.563 x weight in kg) + (1.850 x height in cm) - (4.676 x age in years)] * 1.55] - (weeklyLossInKg * 1000)
+  - For males: calorieGoal = [[66.47 + (13.75 x weight in kg) + (5.003 x height in cm) - (6.755 x age in years)] * 1.55] - (weeklyLossInKg * 1000)
```
Comment (IshaaanVyas) [inline docs/DeveloperGuide.md]: Maybe could explain why these specific formulas are used

**Review by IshaaanVyas — COMMENTED** [review]
Good DG

File: docs/DeveloperGuide.md [hunk 18]
```diff
@@ -1,38 +1,451 @@
...
+1. When the user enters a command, `FitNUS` uses the Parser class to parse the user command.
+2. This results in a `Command` object (more precisely, an object of one of its subclasses e.g., `AddFoodEntryCommand`). 
+3. The `Command` object calls its `execute` method which performs the function required.
+   - Since the `execute` method receives the `FoodDatabase`, `EntryDatabase` and `MealPlanDatabase` initialised in `FitNUS`, it is able to perform operations related to those components (e.g. to add a food tracker entry).
+4. The `execute` method returns a `String` object that contains the outcome message of the command that was executed, which is displayed to the user by the `Ui` component.
+---
+
+### Food Tracker
+
+![tracker class diagram](diagrams/tracker%20class%20diagram.png)
```
Comment (jonathanmui4) [inline docs/DeveloperGuide.md]: Perhaps, you can consider colour coding your classes in the class diagram to make it more readable. You can also consider moving the Gender enum and WeightProgessEntry class into a separate diagram since it does not have any associations/relationships with the rest of the classes in FoodTracker. (Can ignore this part if I'm understanding this wrong 😅)

image

**Review by jonathanmui4 — COMMENTED** [review]
Good job overall! Explanations in the Developer Guide are generally quite clear. Just some formatting issues here and there. Good job! 💪

File: docs/DeveloperGuide.md (hunk 14 above)
Comment (kahhe) [inline docs/DeveloperGuide.md]: A good deal of the hyperlinks are not working on my side as well, only the ` Instruction for manual testing`  works for me.

File: docs/DeveloperGuide.md [hunk 19]
```diff
@@ -1,38 +1,451 @@
...
 
 ### Value proposition
 
-{Describe the value proposition: what problem does it solve?}
+Help user to keep track of their daily calorie intake, and manage their diet wisely.
+
+## Quick Start
+
+1. Ensure you have Java 11 or above installed in your Computer. 
+2. Download the latest fitnus.jar from here (no link for now).
```
Comment (kahhe) [inline docs/DeveloperGuide.md]: Consider having the link to where to download Java `11` will make the DG more useful for the users

File: docs/DeveloperGuide.md [hunk 20]
```diff
@@ -1,38 +1,451 @@
...
 
-{Describe the value proposition: what problem does it solve?}
+Help user to keep track of their daily calorie intake, and manage their diet wisely.
+
+## Quick Start
+
+1. Ensure you have Java 11 or above installed in your Computer. 
+2. Download the latest fitnus.jar from here (no link for now). 
+3. Copy the file to the folder you want to use as the home folder for your WellNUS Tracker 
+4. Type the following command in your terminal to run this program: `java -jar fitnus.jar` (You should change directory to where the `wellnus.jar` file is located or provide the absolute path of `wellnus.jar`).
```
Comment (DJ-Tan) [inline docs/DeveloperGuide.md]: There's 2 namings of .jar file here: 'fitnus.jar' and 'wellnus.jar'

File: docs/DeveloperGuide.md (hunk 9 above)
Comment (kahhe) [reply docs/DeveloperGuide.md:77]: Should the arrow be going towards the user (person out of box) or away? Also for the diagram, do label the unnamed cylinder for greater understandability

File: docs/DeveloperGuide.md (hunk 18 above)
Comment (kahhe) [reply docs/DeveloperGuide.md]: Some visibilities have a space before the method and some don't. Having a space will make it neater in my opinion

File: docs/DeveloperGuide.md [hunk 21]
```diff
@@ -1,38 +1,451 @@
...
-|--------|----------|---------------|------------------|
-|v1.0|new user|see usage instructions|refer to them when I forget how to use the application|
-|v2.0|user|find a to-do item by name|locate a to-do without having to go through the entire list|
+|Version| As a ... | I want to ... | So that I can ...| |--------|----------|---------------|------------------| |v1.0|new user|see usage instructions|refer to them when I forget how to use the application| |v2.0|user|find a to-do item by name|locate a to-do without having to go through the entire list|
+
+## Architecture
+
+### Overall Architecture
+
+![Overall Architecture Diagram](diagrams/overall%20architecture.png)
```
Comment (DJ-Tan) [inline docs/DeveloperGuide.md]: I believe that the arrow for the user should be facing the other way, since the user interacts with the Ui. Also the lines going out pointing to external elements should be dotted? Not very sure but I referenced the AB3 Developer Guide here
image

File: docs/DeveloperGuide.md (hunk 18 above)
Comment (DJ-Tan) [reply docs/DeveloperGuide.md]: For the second variable '-Integer: calories', I believe the team meant it as 'calories: int'. In addition, Integer means something else in Java.
image

File: docs/DeveloperGuide.md [hunk 22]
```diff
@@ -1,38 +1,451 @@
...
+<h4>Implementation</h4>
+<ul>
+<li><h5>Identifying type of method called</h5>
+
+The <code>Parser</code> is invoked through the <code>parseCommandType()</code> method. The input is first split up by identifying a space character.
+If no space character is detected, and the <code>help</code> or <code>exit</code> method was not called, a <code>FitNusException</code> is thrown. The first string element is 
+then compared with default list of commands to determine the type of method called using if-else statements.
+</li></ul></li></ol>
+
+<h2 id="instruction-for-manual-testing"> Instructions for manual testing</h2>
```
Comment (kahhe) [inline docs/DeveloperGuide.md]: There's already a Instructions for manual testing header above.

File: docs/DeveloperGuide.md [hunk 23]
```diff
@@ -1,38 +1,451 @@
...
+![Food Tracker Class Diagram](diagrams/FoodTrackerDatabase_Class.png)
+
+The `FoodTrackerDatabase` component consists of:
+- `addEntry()` Adds a FoodTracker object to the database.
+- `sortDatabase()` Sorts the database by date.
+- `deleteEntry(int)` Removes a specified FoodTracker object from the database.
+- `getTotalDailyCalorie()` Returns the total calorie intake for the day.
+- `convertDatabaseToString()` Returns a String representation of all FoodTracker objects in the database.
+  
+  ![convertDatabaseToString Sequence Diagram](diagrams/EntryDatabase_convertDatabaseToString_Seq.png)
```
Comment (DJ-Tan) [inline docs/DeveloperGuide.md]: For Step 12, I believe the self-invocation should have its own activation bar, and another arrow to return to call.
image

File: docs/DeveloperGuide.md [hunk 24]
```diff
@@ -1,38 +1,451 @@
...
+
+- `getPastMonthsEntryDatabase()` Returns a subset of the original database consisting of FoodTracker objects added in the current month
+
+  ![getPastMonthsEntryDatabase Sequence Diagram](diagrams/EntryDatabase_getPastMonthsEntryDatabase_Seq.png) 
+
+- `editEntryAtIndex(int, Food)` Edits the FoodTracker object at the specified index to the new specified Food object
+
+The diagram below showcases the relationships between FoodTrackerDatabase object and various components.
+
+![Food Tracker Class Architecture](diagrams/FoodTrackerDatabase_Classes.png)
```
Comment (kahhe) [inline docs/DeveloperGuide.md]: image

The implementation has a high degree of coupling with `FoodTrackerDatabase` maybe I'm wrong if it is broken down into other classes that inherit it

File: docs/DeveloperGuide.md [hunk 25]
```diff
@@ -1,38 +1,451 @@
...
+- `getEntryAtIndex(int)` Returns the FoodTracker object at the specified index.
+- `listEntries()` Returns a formatted String of all Food objects to be printed.
+- `findEntries(String)` Returns an ArrayList containing matching FoodTracker objects based on a keyword.
+- `getPastDaysEntryDatabase(int)` Returns a subset of the original database consisting of FoodTracker objects added in the current day
+
+  ![getPastDaysEntryDatabase Sequence Diagram](diagrams/EntryDatabase_getPastDaysEntryDatabase_Seq.png)
+
+- `getPastMonthsEntryDatabase()` Returns a subset of the original database consisting of FoodTracker objects added in the current month
+
+  ![getPastMonthsEntryDatabase Sequence Diagram](diagrams/EntryDatabase_getPastMonthsEntryDatabase_Seq.png)
```
Comment (kahhe) [inline docs/DeveloperGuide.md]: In the two sequence diagrams, self-invocation should generate another activation bar

**Review by kahhe — COMMENTED** [review]
The DG was very detailed and I could tell there was a lot of effort done on it. I think it's also why there are some errors here and there. One ironed out it will look great.
